src/app/clusters/device-energy-management-server/device-energy-management-server.cpp

```diff
@@ -0,0 +1,517 @@
+    Status status        = Status::Success;
+
+    auto powerAdjustmentCapability = mDelegate.GetPowerAdjustmentCapability();
+
+    if (powerAdjustmentCapability.IsNull())
+    {
+        ChipLogError(Zcl, "DEM: %s powerAdjustmentCapability IsNull", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+
+    /* PowerAdjustmentCapability is a list - so iterate through checking if the command is within one of the offers */
+    for (auto pas : powerAdjustmentCapability.Value())
+    {
+        if ((power >= pas.minPower) && (durationSec >= pas.minDuration) && (power <= pas.maxPower) &&
+            (durationSec <= pas.maxDuration))
+        {
+            ChipLogProgress(Zcl, "DEM: %s Good PowerAdjustment args", __FUNCTION__);
+            validArgs = true;
+            break;
+        }
+    }
+
+    if (!validArgs)
+    {
+        ChipLogError(Zcl, "DEM: %s invalid request range", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+
+    ChipLogProgress(Zcl, "DEM: %s Good PowerAdjustRequest() args.", __FUNCTION__);
+
+    /* Call the delegate to do the power adjustment */
+    status = mDelegate.PowerAdjustRequest(power, durationSec);
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+    if (status != Status::Success)
+    {
+        ChipLogError(Zcl, "DEM: %s Failed to PowerAdjustRequest() args.", __FUNCTION__);
+    }
+    return;
+}
+
+void Instance::HandleCancelPowerAdjustRequest(HandlerContext & ctx,
+                                              const Commands::CancelPowerAdjustRequest::DecodableType & commandData)
+{
+    Status status = Status::Success;
+    ESAStateEnum esaStatus;
+
+    /* Check that the ESA state is PowerAdjustActive */
+    esaStatus = mDelegate.GetESAState();
+    if (ESAStateEnum::kPowerAdjustActive != esaStatus)
+    {
+        ChipLogError(Zcl, "DEM: %s - kPowerAdjustActive != esaStatus", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    /* Call the delegate to cancel the power adjustment */
+    status = mDelegate.CancelPowerAdjustRequest();
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+    if (status != Status::Success)
+    {
+        ChipLogError(Zcl, "DEM: %s Failed to CancelPowerAdjustRequest()", __FUNCTION__);
+        return;
+    }
+
+    return;
+}
+
+void Instance::HandleStartTimeAdjustRequest(HandlerContext & ctx,
+                                            const Commands::StartTimeAdjustRequest::DecodableType & commandData)
+{
+    Status status                                               = Status::Success;
+    uint32_t earliestStartTime                                  = 0;
+    uint32_t latestEndTime                                      = 0;
+    DataModel::Nullable<Structs::ForecastStruct::Type> forecast = mDelegate.GetForecast();
```

Comment:
Could probably use auto to save the long type name for forecast.

Reply:
Tried to be consistent in the file, using auto for iterators etc. But for attribute types I feel it is useful to be explicit.